www/js/app.js

```diff
@@ -932,6 +932,12 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
         }
 
         function loadCSSJQuery() {
+            // Ensure all headers are open for clients that lack JavaScript support [kiwix-js #355]
+            var iframe = document.getElementById('articleContent').contentDocument;
+            var collapsedBlocks = iframe.querySelectorAll('.collapsible-block:not(.open-block)');
+            for (var i = collapsedBlocks.length; i--;) {
```

Comment:
@mossroy As there are potentially a large number of blocks in a long document,  I've optimized this loop for speed using a decrement iterator. See https://stackoverflow.com/questions/3520688/javascript-loop-performance-why-is-to-decrement-the-iterator-toward-0-faster-t for explanation of syntax (it's not a mistake!).

Reply:
Interesting!
I've created https://jsperf.com/loop-with-increment-or-decrement-operator to test that on recent browsers. The result really depends on the browser engine.
From what I tested : 
- On Firefox 52.x, the result is the same
- On Firefox OS 2.2 (the simulator, using the Firefox 37 engine), the decrementing loop is 20% faster
- On IE11, the decrementing loop is 30% faster
- But on Chrome 65, the decrementing loop is 20% slower

In any case, I think this is not crucial, as the number of collapsed blocks should not be very high (around 10, I would say).
I suppose that modifying the DOM could be much more expensive than the loop itself.

Feel free to keep the decrementing loop : in this case, please add a comment above with the stackoverflow link.

Regarding the cost of DOM modification, it might be worth optimizing it (in general in the jQuery mode). See section 5 of http://desalasworks.com/article/javascript-performance-techniques/ or section 7 of https://www.sitepoint.com/10-ways-minimize-reflows-improve-performance/ : the browser re-computes the layout in each loop iteration.
Regarding this PR, as the CSS class does not exist yet when we run this loop, maybe it's fast enough. Else jQuery allows to apply this kind of changes in a batch.

I did not run benchmarks on this, and it might not be crucial for this PR. But it might be an idea to optimize our jQuery mode in general

Reply:
Thanks for the ideas. 
I checked here: https://jsperf.com/jquery-addclass-vs-element-classlist-add . 
Test is: jqueryObject.addClass("test"); vs domElement.classList.add("test");
On Edge jQuery is 88% slower.
On Firefox 52, jQuery is 80% slower.
On IE 11, jQuery is 87% slower.
I'll have an explore to see whether jQuery delays the screen repaint but as the for loop is extremely fast, I imagine there would be no time for a full repaint before all the nodes are updated. Anyway, I'll check out what method jQuery uses to mass-update a collection of nodes and will get back.
PS The decrement loop works on all the browsers we support, I believe, but I agree with you that it's probably not a performance bottleneck! I just discovered it and thought I'd try it out.

Reply:
What should be compared is a batch-update with jQuery like (untested) : 
`jQuery('.collapsible-block:not(.open-block)').addClass('open-block');`
From what I understood, this code does all the DOM changes at the same time, triggering only one repaint.

Reply:
I'm not noticing any issue on even large pages such as the Paris page in English Wikivoyage run in FFOS simulator. There are no multiple repaints that are noticeable to the user.

Doing some research on this, I've come across several times the idea that browsers queue reflows and repaints "so if you do four dom modifications in one sequential piece of Javascript, the browser will wait for that piece of Javascript to finish running and then do one reflow and one repaint". So long as your DOM modifications are done in one synchronous piece of JavaScript, no reflow or repaint is triggered until the script stops running. Since we are using synchronous code here and no callback, the update to all the nodes causes a single repaint when the script stops running (or when it makes an asynchronous call). See https://stackoverflow.com/questions/37039667/executing-multiple-dom-updates-with-javascript-efficiently 

Still, if you would prefer to use jQuery here, it's fine by me. All it does is abstract an underlying for or while loop more slowly than doing it ourselves. Underneath jQuery are just standard JavaScript calls and a bunch of syntax parsers that provide the abstraction.

Reply:
Hey @mossroy , I set up a more realistic performance test on real-world data using the two routines we want to compare, here: https://jsperf.com/queryselectorall-classlist-add-vs-jquery/1 

My results are:
Edge 16: jQuery is 16% slower
IE 11: jQuery is 15% slower
Firefox 52: jQuery is 31% slower

Reply:
Actually, use this one:
https://jsperf.com/queryselectorall-classlist-add-vs-jquery-addclass/1 
It has more blocks that match so that the loop runs a few times on each iteration.

Reply:
Alternative to all of this is that I can do a replacement regex that runs on the raw html before we inject it into the DOM. Would be interesting to compare that to the pure JavaScript test...